pkg/il/interpreter/extern.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package interpreter
+
+import (
+	"fmt"
+	"math"
+	"reflect"
+
+	"istio.io/mixer/pkg/il"
+)
+
+// Extern represents an external, native function that is callable from within the interpreter,
+// during program execution.
+type Extern struct {
+	name   string
+	pTypes []il.Type
+	rType  il.Type
+
+	v reflect.Value
+}
+
+// ExternFromFn creates a new reflection based Extern, based on the given function. It panics if the
+// function signature is incompatible to be an extern.
+//
+// A function can be extern under following conditions:
+// - Input parameter types are one of the supported types: string, bool, int64, float64, map[string]string.
```

Comment:
Not yet? We can add support for these. time.Duration is int64, so effectively we can already support it. time.Time seems a bit heavy weight (64bits + 32bits + pointer), which needs to be kept in "heap". Similarly for bytes.